docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+
+## Proposed Design
+
+The main design consideration is where to execute the hooks. At a high level, there are three options:
+
+1. Within the push job.
+2. Within the parent controller.
+3. Within the child controllers.
+
+It is also possible to consider invoking some hooks in one of these location while other hooks would be executed 
+elsewhere. The table below summarizes the feasibility and tradeoffs for each of the proposed hooks:
+
+|                                              | Child Controller | Parent Controller | Push Job |
+|---------------------------------------------:|:----------------:|:-----------------:|:--------:|
+|                          `preStartOfPushJob` |        ❌         |         ✅         |    ✅     |
+|                         `postStartOfPushJob` |        ❌         |         ✅         |    ✅     |
+|                      `preSchemaRegistration` |        ❌         |         ✅         |   1️⃣    |
+|                     `postSchemaRegistration` |        ✅         |         ✅         |   1️⃣    |
+|                    `preStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|                   `postStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|  `preStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+| `postStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+|          `postStoreVersionLeaderReplication` |        ✅         |         ❌         |    ❌     |
+|                        `preStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                       `postStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                            `preEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+|                           `postEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+
+The cells with numbers in them represent special cases, which are explained below:
+
+1️⃣ : Schema registration hooks in push jobs could only be invoked in cases where auto-registration is enabled and the 
+new schema originates from the push job itself, whereas schema registrations which are performed directly on the 
+controller could not trigger the hook.
+
+2️⃣ : The hook for the start of ingestion for Da Vinci Clients is tricky for a few reasons. The start of ingestion is 
+controlled by updating the Meta Store, which is a non-replicated system store updated by child controllers, so those
+must be involved (either by running the hook there in the first place, or by having some mechanism that enables the
+parent or VPJ to inform the child controllers of when the system store is eligible for getting updated, such as by
+adding a new field to the `AddVersion` admin channel command). However, see Rollback Support below for why running hooks
+in the child controller may be insufficient.
+
+### Rollback Support
```

Comment:
So the rollback option only makes sense in `postStoreVersionSwap` stage? for previous steps what does rollback mean?

Reply:
If a hook returns `ROLLBACK` prior to the first region having swapped, then it's simply equivalent to `ABORT`... but if any other region already swapped, then it can be done.

The main purpose of the `preStoreVersionSwap` is for the option to return `WAIT`.

If we want, we could have three enums instead of two... and the additional enum would contain `PROCEED`, `ABORT` and `WAIT` but not `ROLLBACK`... but it seems a bit tedious to have this many enums... the API is simpler with fewer enums, IMHO.

But feel free to disagree, of course. LMK what you think.